node/src/daser.rs

```diff
@@ -110,74 +125,295 @@ where
+        });
+
         Ok(())
     }
 
-    async fn sample_block(&mut self, header: &ExtendedHeader) -> Result<(Vec<Cid>, bool)> {
-        let now = Instant::now();
-        let indexes = random_indexes(header.dah.square_width(), self.max_samples_needed);
-        let mut futs = FuturesUnordered::new();
+    /// Rollback to the highest accepted block.
+    ///
+    /// Rollback is needed because of how Shwap works: The only way to know
+    /// if sampling has failed is via a timeout on bitswap requests and this
+    /// can be generated from network issues too. We rollback just to recheck
+    /// the previous failed samplings.
+    async fn rollback(&mut self) {
+        // TODO
+    }
 
-        for (row_index, column_index) in indexes {
-            let fut = self
-                .p2p
-                .get_sample(row_index, column_index, header.height().value());
-            futs.push(fut);
-        }
+    async fn populate_queue(&mut self) {
+        // TODO: Adjust algorithm for backward syncing.
 
-        let mut cids: Vec<Cid> = Vec::new();
-        let mut accepted = true;
-
-        while let Some(res) = futs.next().await {
-            match res {
-                Ok(sample) => cids.push(convert_cid(&sample.id.into())?),
-                // Validation is done at Bitswap level, through `ShwapMultihasher`.
-                // If the sample is not valid, it will never be delivered to us
-                // as the data of the CID. Because of that, the only signal
-                // that data sampling verification failed is query timing out.
-                Err(P2pError::BitswapQueryTimeout) => accepted = false,
-                Err(e) => return Err(e.into()),
+        let Ok(new_head) = self.store.head_height().await else {
+            return;
+        };
+
+        match self.prev_head {
+            Some(prev_head) => {
+                for height in prev_head + 1..=new_head {
+                    if is_sampled(&*self.store, height).await {
+                        continue;
+                    }
+
+                    let Ok(header) = self.store.get_by_height(height).await else {
+                        // Maybe header was pruned, continue to the next one.
+                        continue;
+                    };
+
+                    if !in_sampling_window(header.time()) {
+                        // Block is older than the sampling window, continue to the next one.
+                        continue;
+                    }
+
+                    queue_sampling(&mut self.queue, &header);
+                }
+            }
+            None => {
+                // Reverse iterate heights
+                for height in (1..=new_head).rev() {
+                    if is_sampled(&*self.store, height).await {
+                        continue;
+                    }
+
+                    let Ok(header) = self.store.get_by_height(height).await else {
+                        // We reached the tail of the known heights
```

Comment:
Mostly to confirm, 
> TODO: Adjust algorithm for backward syncing

above is to account for upcoming backwards header sync changes, that make it so that we're no longer guaranteed to store single continuous block of headers in store?

Reply:
Yes. Now the algorithm iterates from head to block 1. This is not possible with backward syncing.